cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "scan.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/scan.h>
+#include <thrust/transform_scan.h>
+
+namespace cudf {
+namespace detail {
+
+template <typename T>
+using pair_type = thrust::pair<T, T>;
+
+/**
+ * @brief functor to be summed over in a prefix sum such that
+ * the recurrence in question is solved.
+ * see https://www.cs.cmu.edu/~guyb/papers/Ble93.pdf S. 1.4
+ * for details
+ */
+template <typename T>
+class recurrence_functor {
+ public:
+  __device__ pair_type<T> operator()(pair_type<T> ci, pair_type<T> cj)
+  {
+    return {ci.first * cj.first, ci.second * cj.first + cj.second};
+  }
+};
+
+template <typename T>
+class ewma_functor_base {
+ public:
+  pair_type<T> IDENTITY = {1.0, 0.0};
+};
+
+template <typename T>
+class ewma_adjust_numerator_nonull_functor : public ewma_functor_base<T> {
```

Comment:
these names don't seem great, what should I call these?

Reply:
Before addressing naming, let's see if there's a better approach here. I think you can implement all of these functors as a single functor with a couple of additional template parameters and suitable use of `if constexpr`. Could you give that a shot? For example, you could do things like
```
// We may want to make has_nulls a runtime nullate rather than a template parameter, but for now let's just try it with templating since that's easier.
template <typename T, bool adjust, bool has_nulls, bool is_numerator>
...
auto input = [=] () {
    if constexpr (is_numerator) {
        return thrust::get<2>(data);
    } else {
        return 1.0;
    }
}();
...
```
That's specifically an example of using the template parameter to make a compile-time decision on what the second value of the pair should be based on whether we're computing the numerator or the denominator, but I think you could do similar things for handling adjustment and nullability. I'm not sure that you'll be able to combine _all_ of the functors that way, and some of them may just be clearer/more readable if you leave them with separate implementations, but right now there's a lot of code duplication and the naming confusion is partly a result of that. Let's try and use templates to resolve this at least part way.